### PR TITLE
Receive strategy refactoring

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Connects to https://github.com/Particular/NServiceBus.SqlServer/issues/232

Refactored receive strategies in order to reduce the complexity and (most importantly) code quadruplication.

 * Introduce integration tests that verify the contract for a receive strategy
 * Introduce dispatch strategy as a counterpart on send side
 * Simplified communication between receive and send strategies (introducing `IDispatchPolicy`)
 * Moved the responsibility of managing *no more messages inside the loop* case to where it belongs (message pump)
 * Factored out the common algorithm shared by all receive strategies and collapsed individual implementations to a single factory method

Please review commit-by-commit because history is important here.

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@Particular/sqlserver-transport-maintainers please review.

File: src/NServiceBus.SqlServer.IntegrationTests/ReceiveStrategyTests.cs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+            creator = new QueueCreator(sqlConnectionFactory, addressParser);
+            queuePurger = new QueuePurger(sqlConnectionFactory);
+
+            await creator.CreateQueueIfNecessary(queueBindings, "");
+            await queuePurger.Purge(queue);
+            await queuePurger.Purge(errorQueue);
+        }
+
+        TableBasedQueue queue;
+        TableBasedQueue errorQueue;
```
Comment (HEskandari) [inline src/NServiceBus.SqlServer.IntegrationTests/ReceiveStrategyTests.cs]: Couldn't spot a cleanup method. Do we need one to ensure failing of one test doesn't have domino effect on the other tests? e.g. if one test fails and leaves a message in the queue.

**Comment by HEskandari** [comment]
Seemed fine to my eyes. could you please fix the conflict @SzymonPobiega ?

File: src/NServiceBus.SqlServer.UnitTests/Sending/MessageDispatcherTests.cs [hunk 2]
```diff
@@ -69,17 +69,31 @@ static TransportOperation CreateTransportOperations(string messageId, string des
...
-            public Task DispatchAsNonIsolated(List<MessageWithAddress> operations, ContextBag context)
+            public IDispatchStrategy CreateNonIsolatedDispatchStrategy()
             {
-                DispatchedMessageIds.AddRange(operations.Select(x => x.Message.MessageId));
-                return Task.FromResult(0);
+                return new DispatchStrategy(DispatchedMessageIds);
             }
 
-            public Task DispatchAsIsolated(List<MessageWithAddress> operations)
+            class DispatchStrategy : IDispatchStrategy
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer.UnitTests/Sending/MessageDispatcherTests.cs]: Can we call this `FakeDispatchStrategy` or something similar to indicate it is not the real deal?

File: src/NServiceBus.SqlServer/Receiving/ReceiveStrategyContext.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    using System.Data.SqlClient;
+    using System.Threading.Tasks;
+    using System.Transactions;
+    using JetBrains.Annotations;
+
+    class ReceiveStrategyContext : IDisposable
+    {
+        public ReceiveStrategyContext([NotNull] SqlConnection connection, [NotNull] IDispatchStrategy dispatchStrategy)
+        {
+            if (connection == null)
```
Comment (MarcinHoppe) [inline src/NServiceBus.SqlServer/Receiving/ReceiveStrategyContext.cs]: Use `Guard`?

**Comment by MarcinHoppe** [comment]
@SzymonPobiega I posted some minor comments:

https://github.com/Particular/NServiceBus.SqlServer/pull/227#discussion_r62804717
https://github.com/Particular/NServiceBus.SqlServer/pull/227#discussion_r62806696

Otherwise LGTM. I like the change a lot.

:beers:

**Comment by SzymonPobiega** [comment]
@MarcinHoppe I've implemented your changes. I actually decided to remove the null checks because these classes are internal.

File: src/NServiceBus.SqlServer.IntegrationTests/ReceiveStrategyTests.cs (hunk 1 above)
Comment (SzymonPobiega) [inline src/NServiceBus.SqlServer.IntegrationTests/ReceiveStrategyTests.cs]: The test uses specific queues (`ReceiveStrategyTests` and `ReceiveStrategyTests.DLQ`) and it cleans these queues up in the `Setup` method so we should be fine.

**Comment by SzymonPobiega** [comment]
@tmasternak closing this one, what do you think?

**Comment by tmasternak** [comment]
@SzymonPobiega yep I agree

**Pull request closed without merging**